src/oracle/middleware.js

```diff
@@ -0,0 +1,761 @@
+  */
+export const getOffers = async (follower) => {
+
+  let history = [];
+  let counter = 0;
+  let lastVisited = 0;
+
+  for await (const followerElement of iterateReverse(follower)) {
+
+    if (counter == 5){
+      break;
+    }
+
+    //if it is an offer status 
+    if (followerElement.value.updated == "offerStatus"){
+      //get id
+      let id = followerElement.value.status.id
+
+      //if a new and final state
+      if (id != lastVisited) {
+        //if it is not failed
+        if (!followerElement.value.status.hasOwnProperty("error")) {
+          history.push(followerElement.value);
+          counter++
+        }
+        lastVisited = id;
+      } 
+    }
+  }
+  return history
+}
+
+/**
+ * Function to get the latest submitted round
+ * @returns the latest round submitted
+ */
+export const getLatestSubmittedRound = async () => {
+  const unserializer = boardSlottingMarshaller(fromBoard.convertSlotToVal);
+  const leader = makeLeader(networkConfig.rpcAddrs[0]);
+
+  const follower = await makeFollower(
+    `:published.wallet.${FROM}`,
+    leader,
+    {
+      // @ts-expect-error xxx
+      unserializer,
+    },
+  );
+
+  //get offers
+  let offers = await getOffers(follower)
+
+  return Number(offers[0]["status"]["invitationSpec"]["invitationArgs"][0]["roundId"])
+}
+
+/**
+ * Function to check if submission was satisfied for a specific round
+ * @param {*} oracle oracle address
+ * @param {*} feedOfferId the offer id of the feed to check for
+ * @param {*} roundId the round Id which is checked
+ * @returns whether a submission was successful for a specific round
+ */
+export const checkSubmissionForRound = async (oracle, feedOfferId, roundId) => {
+  const unserializer = boardSlottingMarshaller(fromBoard.convertSlotToVal);
+  const leader = makeLeader(networkConfig.rpcAddrs[0]);
+
+  const follower = await makeFollower(
+    `:published.wallet.${oracle}`,
+    leader,
+    {
+      // @ts-expect-error xxx
+      unserializer,
+    },
+  );
+
+  //get offers
+  let offers = await getOffers(follower)
+  
+  //loop through offers starting from last offer
+  for (var i = 0; i < offers.length; i++) {
+    //get current offer
+    var currentOffer = offers[i];
+
+    //if a price invitation and for the correct feed
+    if (currentOffer["status"]["invitationSpec"]["invitationMakerName"] == "PushPrice" && currentOffer["status"]["invitationSpec"]["previousOffer"] == feedOfferId) {
+      let offerRound = Number(currentOffer["status"]["invitationSpec"]["invitationArgs"][0]["roundId"])
+
+      //if it is an offer for the round we are checking for
+      if (offerRound == roundId) {
+        //if there is no error
+        if (!currentOffer["status"].hasOwnProperty("error")) {
+          return true
+        }
+      }
+      //else if offer round id is less than the round we want to check and its satisfied
+      //return false because there cannot be a submission for a newer round before this offer 
```

Comment:
//else if offer round id is less than the round we want to check and its satisfied
      //return false because there cannot be a submission for a newer round before this offer 
      
      Not a fan of this comment, how about 
      If the currentOffer has no error and the offerRound is less than the roundId, we return false since no new submissions can be created before completing the previous offer